### PR TITLE
Update FFI & Access related code in OpenJ9 for compilation in JDK20

### PR DESCRIPTION
**Pull request by ChengJin01**
The change is to update all FFI & `java.lang.Access` related java code in JDK20
for the moment to get them compiled as we have been working on the changes
to address the  latest APIs intended for JEP434 in JDK20 as follows:

Commit 1: JDK20 java.lang.Access new API
1) added `Access.in()` to support `Access.initialSystemIn()`
2) added the stub implementation of `Java_java_lang_Thread_ensureMaterializedForStackWalk`

Commit 2: Disable FFI related code for JDK20

Signed-off-by: ChengJin01 <jincheng@ca.ibm.com>

### CODE REVIEW

**Comment by ChengJin01** [comment]
The changes are made on top of Jason's changes at https://github.com/eclipse-openj9/openj9/pull/16574, which simply disabled all FFI specific code for JDK20.

Note:
1) the change is only used for compilation. Further update will be made to address the latest APIs in JDK20.
2) the changes must be merged altogether with the changes in OpenJ9 at https://github.com/eclipse-openj9/openj9/pull/16585

**Comment by ChengJin01** [comment]
Reviewer: @tajila, @pshipton
FYI: @JasonFengJ9

**Comment by tajila** [comment]
FYI @babsingh

**Comment by ChengJin01** [comment]
>     * Even though we want to merge all changes together in the same PR, we should keep @JasonFengJ9's changes ([JDK20 java.lang.Access new APIs #16574](https://github.com/eclipse-openj9/openj9/pull/16574)) and FFI changes in separate commits. This will segregate the changes, make it easier to review and easier to revert. Also, cherry-picking Jason's commit from [JDK20 java.lang.Access new APIs #16574](https://github.com/eclipse-openj9/openj9/pull/16574) will preserve his authorship.
> 

Agreed. I will need to close this PR and create a new branch on top of Jason's branch to set things up.

>     * We should also point to the Github issue which will re-enable all the disabled FFI code. If there is none, we should create a Github issue to track this work.

The code we are currently working on is targeted for JEP434 which we already have the issue at https://github.com/eclipse-openj9/openj9/issues/16329 to track the work after that.

>     * Were the compilation failures, which have forced us to disable our FFI code, documented in a Github issue? We should also refer to it in this issue.

The temporary changes is only for the compilation in the `open9-staging` branch of JDK20 mentioned at https://github.com/ibmruntimes/openj9-openjdk-jdk20/pull/9, in which case I don't think any issue is suitable for the temp changes.

**Comment by ChengJin01** [comment]
As talked to @babsingh offline, we don't have to create a new issue for this PR but we will mention these temp changes in https://github.com/eclipse-openj9/openj9/issues/16329 to keep track of the working being done.

**Comment by ChengJin01** [comment]
Just updated the branch with two separate commits as mentioned in the description.

File: runtime/jcl/exports.cmake [hunk 1]
```diff
@@ -688,3 +688,10 @@ if(NOT JAVA_SPEC_VERSION LESS 19)
 		Java_jdk_internal_vm_Continuation_unpin
 	)
 endif()
+
+# java 20+
```
Comment (babsingh) [inline runtime/jcl/exports.cmake:692]:
```suggestion
# Java 20+
```

File: runtime/jcl/common/thread.cpp [hunk 2]
```diff
@@ -571,4 +571,14 @@ Java_java_lang_Thread_registerNatives(JNIEnv *env, jclass clazz)
 }
 #endif /* JAVA_SPEC_VERSION >= 19 */
 
+#if JAVA_SPEC_VERSION >= 20
+
+void JNICALL
+Java_java_lang_Thread_ensureMaterializedForStackWalk(JNIEnv *env, jclass threadClass, jobject obj)
+{
+	/* Stub method, to be implemented */
```
Comment (babsingh) [inline runtime/jcl/common/thread.cpp:579]:
```suggestion
	/* Stub method, to be implemented. */
```

File: test/functional/Java19andUp/playlist.xml [hunk 3]
```diff
@@ -70,7 +70,7 @@
 			-excludegroups $(DEFAULT_EXCLUDE); \
 			$(TEST_STATUS)
 		</command>
-		<platformRequirements>bits.64,^arch.x86,^arch.arm,^arch.riscv,^os.zos,^os.sunos</platformRequirements>
+		<platformRequirements>bits.64,^arch.arm,^arch.riscv,^os.zos,^os.sunos</platformRequirements>
```
Comment (babsingh) [inline test/functional/Java19andUp/playlist.xml:73]: Why are we enabling the test for JDK19 arch.x86 while disabling the test? Another instance seen on line 102.

**Review by babsingh — DISMISSED** [review]
General feedback: The new functions should have a function description. Example: https://github.com/eclipse/omr/blob/master/doc/CodingStandard.md#comment-function-parameters-clearly.

File: test/functional/Java19andUp/playlist.xml (hunk 3 above)
Comment (ChengJin01) [reply test/functional/Java19andUp/playlist.xml:73]: The `x86` was previously disabled for downcall & upcall issue at https://github.com/eclipse-openj9/openj9/issues/16161 which is now fixed by JIT via https://github.com/eclipse-openj9/openj9/pull/16556, in which case we need to re-enable in JDK19.

In addition, all these test suites are only intended for the FFI APIs in JDK19 which will be replaced with the latest APIs in JDK20. So we will create new test suites with the latest APIs in JDK20 once our code is ready to work.

File: runtime/jcl/exports.cmake (hunk 1 above)
Comment (ChengJin01) [reply runtime/jcl/exports.cmake]: Fixed.

File: runtime/jcl/common/thread.cpp (hunk 2 above)
Comment (ChengJin01) [reply runtime/jcl/common/thread.cpp]: Fixed.

**Comment by ChengJin01** [comment]
> General feedback: The new functions should have a function description. Example: https://github.com/eclipse/omr/blob/master/doc/CodingStandard.md#comment-function-parameters-clearly.

@JasonFengJ9, could you provide a description of the new function added to your changes?  It belongs to JEP 429: Implementation of Scoped Values (Incubator) which I didn't find any explanation of this function in JDK20.

**Comment by JasonFengJ9** [comment]
Updated https://github.com/eclipse-openj9/openj9/pull/16574, @ChengJin01 pls pull them in.

**Comment by ChengJin01** [comment]
> Updated #16574, @ChengJin01 pls pull them in.

@JasonFengJ9, I rebased your code after resolving the conflicts. Please help double-check to see whether everything is included from your changes without anything wrong/missing.

File: runtime/jcl/exports.cmake [hunk 4]
```diff
@@ -688,3 +688,10 @@ if(NOT JAVA_SPEC_VERSION LESS 19)
 		Java_jdk_internal_vm_Continuation_unpin
 	)
 endif()
+
+# Java 20+
```
Comment (JasonFengJ9) [inline runtime/jcl/exports.cmake:692]: nit: it is `# java 20+` in https://github.com/eclipse-openj9/openj9/pull/16574

Comment (ChengJin01) [reply runtime/jcl/exports.cmake:692]: This is updated to `'Java 20+` as requested at https://github.com/eclipse-openj9/openj9/pull/16585#discussion_r1082847896.

Comment (ChengJin01) [reply runtime/jcl/exports.cmake:692]: I should update this with `java 20+` to keep consistent with the rest of the script in terms of java version.

Comment (ChengJin01) [reply runtime/jcl/exports.cmake]: Restored to `java 20+`.

Comment (babsingh) [reply runtime/jcl/exports.cmake]: `Java` should be capitalized since it is a name. The incorrect lower-case instances should be fixed; but, they do not need to be fixed in this PR.

**Comment by babsingh** [comment]
Now, there are 4 commits in the PR. The second and fourth commits have code from resolving conflicts, which should be removed. Example:
```
<<<<<<< Upstream, based on upstream/jdk20thread
	static InputStream initialSystemIn;
=======
	static InputStream in;
>>>>>>> 6c36088 JDK20 java.lang.Access new APIs
```

**Comment by ChengJin01** [comment]
> Now, there are 4 commits in the PR. The second and fourth commits have code from resolving conflicts, which should be removed. Example:
> 
> ```
> <<<<<<< Upstream, based on upstream/jdk20thread
> 	static InputStream initialSystemIn;
> =======
> 	static InputStream in;
> >>>>>>> 6c36088 JDK20 java.lang.Access new APIs
> ```

I just re-pulled the Jason's latest changes and placed all my changes on top for that without conflicts, in which case we only keep  two commits.

File: runtime/jcl/exports.cmake (hunk 4 above)
Comment (ChengJin01) [reply runtime/jcl/exports.cmake:692]: Updated with `Java 20` in my changes at https://github.com/eclipse-openj9/openj9/pull/16585/commits/16c03d508fba1962af68b6c713aaa938720e60cf#diff-8e3f29d0e205cd3f6cfd36ef0b25a9a1d2df93d47a86edf81fdf27763b97cb13.

**Comment by babsingh** [comment]
Away during day this week. Passing over the review to @tajila and @pshipton. Ref: https://github.com/eclipse-openj9/openj9/pull/16585#issuecomment-1397934785.

File: test/functional/Java17andUp/playlist.xml [hunk 5]
```diff
@@ -23,13 +23,6 @@
 <playlist xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../../TestConfig/playlist.xsd">
 	<test>
 		<testCaseName>Jep389Tests_testClinkerFfi_DownCall</testCaseName>
-		<disables>
-			<disable>
-				<comment>https://github.com/eclipse-openj9/openj9/issues/14483#issuecomment-1039546787</comment>
```
Comment (pshipton) [inline test/functional/Java17andUp/playlist.xml:28]: Does this change to unexclude the jdk17 test rely on anything else in this PR? It seems unrelated and should be done in a separate PR.

File: test/functional/Java19andUp/playlist.xml (hunk 3 above)
Comment (pshipton) [inline test/functional/Java19andUp/playlist.xml:73]: Pls make this change in a separate PR.

Comment (pshipton) [reply test/functional/Java19andUp/playlist.xml:73]: Pls make this change in a separate PR.

File: test/functional/Java17andUp/playlist.xml (hunk 5 above)
Comment (ChengJin01) [reply test/functional/Java17andUp/playlist.xml]: Agreed and moved it to https://github.com/eclipse-openj9/openj9/pull/16599 given it is related to https://github.com/eclipse-openj9/openj9/issues/14483 (the OpenJDK MH.invokeBasic() which was solved in JIT).

File: test/functional/Java19andUp/playlist.xml [hunk 6]
```diff
@@ -99,7 +99,7 @@
...
```
[the remark this answers is not in the document]
Comment (ChengJin01) [reply test/functional/Java19andUp/playlist.xml]: Moved it to https://github.com/eclipse-openj9/openj9/pull/16599

File: test/functional/Java19andUp/playlist.xml (hunk 3 above)
Comment (ChengJin01) [reply test/functional/Java19andUp/playlist.xml]: Moved it to https://github.com/eclipse-openj9/openj9/pull/16599.

**Review by tajila — APPROVED** [review]

**Comment by tajila** [comment]
is there any testing we can do here @pshipton

File: runtime/jcl/common/thread.cpp [hunk 7]
```diff
@@ -571,4 +571,21 @@ Java_java_lang_Thread_registerNatives(JNIEnv *env, jclass clazz)
...
 #endif /* JAVA_SPEC_VERSION >= 19 */
 
+#if JAVA_SPEC_VERSION >= 20
+
+/* static native void ensureMaterializedForStackWalk(Object o);
+ *
+ * This is expected to invoke JVM_EnsureMaterializedForStackWalk and ensure
+ * that the stackwalk code sees a materialized value.
+ * This might be a hotspot specific requirement, i.e., an empty
+ * implementation could be sufficient for OpenJ9.
```
Comment (babsingh) [inline runtime/jcl/common/thread.cpp:581]: Why params and return value not included in the function description?

Ref: https://github.com/eclipse/omr/blob/master/doc/CodingStandard.md#comment-function-parameters-clearly

File: jcl/src/java.base/share/classes/java/lang/Access.java [hunk 8]
```diff
@@ -543,6 +560,35 @@ public void setContinuation(Thread thread, Continuation c) {
...
+		Thread.setScopedValueBindings(scopeValueBindings);
+	}
+
+	public void ensureMaterializedForStackWalk(Object obj) {
+		Thread.ensureMaterializedForStackWalk(obj);
+	}
+
+	public InputStream initialSystemIn() {
+		return initialSystemIn;
+	}
```
Comment (babsingh) [inline jcl/src/java.base/share/classes/java/lang/Access.java:590]: Why a function description was not added for any of the Java methods?

**Review by babsingh — COMMENTED** [review]
Missing space in the commit message title: `[FFI]Disable FFI related code for JDK20`.

**Comment by babsingh** [comment]
Preliminary testing:

jenkins test sanity zlinux jdk20,jdk11

**Comment by babsingh** [comment]
re https://github.com/eclipse-openj9/openj9/pull/16585#issuecomment-1402433044: this build does not run against jdk20#openj9-staging. looking for the cmd to run a build against jdk20#openj9-staging.

**Comment by pshipton** [comment]
jenkins test sanity,sanity.openjdk all jdk20 depends ibmruntimes/openj9-openjdk-jdk20#9

File: runtime/jcl/common/thread.cpp (hunk 7 above)
Comment (ChengJin01) [reply runtime/jcl/common/thread.cpp:581]: FYI: @JasonFengJ9

File: jcl/src/java.base/share/classes/java/lang/Access.java (hunk 8 above)
Comment (ChengJin01) [reply jcl/src/java.base/share/classes/java/lang/Access.java:590]: FYI: @JasonFengJ9

**Comment by pshipton** [comment]
Actually it's

jenkins test sanity,sanity.openjdk all jdk20 depends https://github.com/ibmruntimes/openj9-openjdk-jdk20/pull/12

File: jcl/src/java.base/share/classes/java/lang/Access.java (hunk 8 above)
Comment (tajila) [reply jcl/src/java.base/share/classes/java/lang/Access.java:590]: We typically dont document these because they just forward to other methods which are documented. Duplicating comments gets out of sync the moment we need to update something

File: runtime/jcl/common/thread.cpp (hunk 7 above)
Comment (JasonFengJ9) [reply runtime/jcl/common/thread.cpp:581]: @ChengJin01 you might pull the updated PR w/ updated function doc
* https://github.com/eclipse-openj9/openj9/pull/16574

File: jcl/src/java.base/share/classes/java/lang/Access.java (hunk 8 above)
Comment (JasonFengJ9) [reply jcl/src/java.base/share/classes/java/lang/Access.java:590]: Yeah, `Access` is a helper class to allow privileged access to classes from outside `java.lang` package in which most of the methods just forward to other methods. Added java doc for `initialSystemIn()` which returns the initial system `in` set by `System.setIn()` instead of invoking another method.

File: runtime/jcl/common/thread.cpp (hunk 7 above)
Comment (ChengJin01) [reply runtime/jcl/common/thread.cpp:581]: @JasonFengJ9, could you help update `java 20+`  to `Java 20+` in https://github.com/eclipse-openj9/openj9/pull/16574 as requested by Babneet at https://github.com/eclipse-openj9/openj9/pull/16585#discussion_r1083034945 ? so I don't need to update again in my own commit.

Comment (JasonFengJ9) [reply runtime/jcl/common/thread.cpp:581]: Done

Comment (ChengJin01) [reply runtime/jcl/common/thread.cpp:581]: Updated the changes in the latest commits.

**Comment by pshipton** [comment]
Pls update the PR title and description here as well, to include mention of the updated APIs added for jdk20 (more than just listing the commit).

**Comment by ChengJin01** [comment]
> Pls update the PR title and description here as well, to include mention of the updated APIs added for jdk20 (more than just listing the commit).

The PR tile & description are updated as suggested above.

**Comment by babsingh** [comment]
> https://openj9-jenkins.osuosl.org/job/PullRequest-OpenJ9/3249/

In the PR builds, compilation errors and test failures are seen while running the openjdk.sanity test suite. @ChengJin01 What's the plan to handle them (e.g. document, disable and fix)?

**Comment by ChengJin01** [comment]
> > https://openj9-jenkins.osuosl.org/job/PullRequest-OpenJ9/3249/
> 
> In the PR builds, compilation errors and test failures are seen while running the openjdk.sanity test suite. @ChengJin01 What's the plan to handle them (e.g. document, disable and fix)?

I've gone over these builds which only failed in tests (all builds passed in compilation) e.g.https://openj9-jenkins.osuosl.org/job/Test_openjdk20_j9_sanity.openjdk_x86-64_mac_Personal/1/tapResults/  (which has nothing to do with FFI)
```
Failed test cases: 
TEST: java/lang/Class/getEnumConstants/BadEnumTest.java
TEST: java/lang/invoke/VarHandles/VarHandleTestExact.java
TEST: java/lang/reflect/Proxy/ProxyModuleMapping.java
TEST: sun/invoke/util/ValueConversionsTest.java
Test results: passed: 838; failed: 5
```
I think they might need to be excluded accordingly for now.

FYI: @JasonFengJ9

**Comment by ChengJin01** [comment]
@JasonFengJ9, it would be better to create issues for these failing test suites separately to track them for later fix/exclude.

**Comment by pshipton** [comment]
@ChengJin01 your ECA check failed, not sure why. Maybe just a fluke, when I checked your email manually it passed. I guess we'll ignore it.

**Comment by ChengJin01** [comment]
> @ChengJin01 your ECA check failed, not sure why. Maybe just a fluke, when I checked your email manually it passed. I guess we'll ignore it.

I can log in with my e-mail at https://accounts.eclipse.org/users/cjinqv0 without any issue.

**Comment by babsingh** [comment]
re https://github.com/eclipse-openj9/openj9/pull/16585#issuecomment-1402872890:
> all builds passed in compilation
> https://openj9-jenkins.osuosl.org/job/Test_openjdk20_j9_sanity.openjdk_x86-64_mac_Personal/1/consoleFull

Compilation failures show up in the full console log (grep for `error:`). There are many instances of the below error.
```
15:02:35  /Users/jenkins/workspace/Test_openjdk20_j9_sanity.openjdk_x86-64_mac_Personal/aqa-tests/openjdk/openjdk-jdk/test/jdk/java/lang/invoke/VarHandles/VarHandleTestExact.java:51: error: cannot find symbol
15:02:35  import java.lang.foreign.MemorySession;
15:02:35                          ^
15:02:35    symbol:   class MemorySession
15:02:35    location: package java.lang.foreign
15:02:35  /Users/jenkins/workspace/Test_openjdk20_j9_sanity.openjdk_x86-64_mac_Personal/aqa-tests/openjdk/openjdk-jdk/test/jdk/java/lang/invoke/VarHandles/VarHandleTestExact.java:173: error: cannot find symbol
15:02:35          try (MemorySession session = MemorySession.openConfined()) {
15:02:35               ^
15:02:35    symbol:   class MemorySession
15:02:35    location: class VarHandleTestExact
15:02:35  /Users/jenkins/workspace/Test_openjdk20_j9_sanity.openjdk_x86-64_mac_Personal/aqa-tests/openjdk/openjdk-jdk/test/jdk/java/lang/invoke/VarHandles/VarHandleTestExact.java:173: error: cannot find symbol
15:02:35          try (MemorySession session = MemorySession.openConfined()) {
15:02:35                                       ^
15:02:35    symbol:   variable MemorySession
15:02:35    location: class VarHandleTestExact
15:02:35  Note: /Users/jenkins/workspace/Test_openjdk20_j9_sanity.openjdk_x86-64_mac_Personal/aqa-tests/openjdk/openjdk-jdk/test/jdk/java/lang/invoke/VarHandles/VarHandleTestExact.java uses preview features of Java SE 20.
15:02:35  Note: Recompile with -Xlint:preview for details.
15:02:35  3 errors
```

**Comment by ChengJin01** [comment]
> MemorySession

This is because the builds picked up the wrong test suites at https://github.com/ibmruntimes/openj9-openjdk-jdk20/blob/openj9/test/jdk/java/lang/invoke/VarHandles/VarHandleTestExact.java rather than https://github.com/ibmruntimes/openj9-openjdk-jdk20/blob/master/test/jdk/java/lang/invoke/VarHandles/VarHandleTestExact.java (in which case `MemorySession` has been totally removed).  That being said, all the OpenJDK test suites with the outdated APIs are wrongly selected (they should choose the test suite in https://github.com/ibmruntimes/openj9-openjdk-jdk20/pull/12 rather than the existing Jtreg test suites in `openj9` branch).

**Comment by pshipton** [comment]
We'll get this stuff merged, see about moving up to the latest jdk-20 build, and then can assess what we need to do for the openjdk tests.

**Review by babsingh — APPROVED** [review]

**Comment by JasonFengJ9** [comment]
> it would be better to create issues for these failing test suites separately to track them for later fix/exclude.

We will see if these failures still happen after merging this PR.

File: jcl/src/java.base/share/classes/java/lang/Access.java [hunk 9]
```diff
@@ -543,6 +560,40 @@ public void setContinuation(Thread thread, Continuation c) {
...
+
+	public void ensureMaterializedForStackWalk(Object obj) {
+		Thread.ensureMaterializedForStackWalk(obj);
+	}
+
+	/**
+	 * Returns the initial system in set by System.setIn().
+	 *
+	 * @return the initialSystemIn
+	 */
```
Comment (keithc-ca) [inline jcl/src/java.base/share/classes/java/lang/Access.java:592]: As @tajila commented, this isn't really helpful. Please remove it for consistency with other methods here.

File: jcl/src/java.base/share/classes/jdk/internal/foreign/abi/DowncallLinker.java [hunk 10]
```diff
@@ -1,6 +1,6 @@
-/*[INCLUDE-IF JAVA_SPEC_VERSION >= 19]*/
+/*[INCLUDE-IF JAVA_SPEC_VERSION == 19]*/
```
Comment (keithc-ca) [inline jcl/src/java.base/share/classes/jdk/internal/foreign/abi/DowncallLinker.java:1]: Is there an issue to track restoring this for jdk20+?

File: runtime/jcl/common/thread.cpp [hunk 11]
```diff
@@ -571,4 +571,27 @@ Java_java_lang_Thread_registerNatives(JNIEnv *env, jclass clazz)
...
+ * @param env instance of JNIEnv
+ * @param unusedClass
+ * @param obj the object to be checked
+ *
+ * @return void
+ */
+void JNICALL
+Java_java_lang_Thread_ensureMaterializedForStackWalk(JNIEnv *env, jclass unusedClass, jobject obj)
+{
+	/* Stub method, to be implemented. */
```
Comment (keithc-ca) [inline runtime/jcl/common/thread.cpp:592]: We may not need to do anything here except, perhaps, update the comment; see [jvm.h](https://github.com/ibmruntimes/openj9-openjdk-jdk20/blob/openj9/src/hotspot/share/include/jvm.h#L743).

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: jcl/src/java.base/share/classes/jdk/internal/foreign/abi/DowncallLinker.java (hunk 10 above)
Comment (ChengJin01) [reply jcl/src/java.base/share/classes/jdk/internal/foreign/abi/DowncallLinker.java:1]: As mentioned previously, we already have the issue at https://github.com/eclipse-openj9/openj9/issues/16329 for JEP3434/JDK20 to update all FFI related code in JDK20+.

File: jcl/src/java.base/share/classes/java/lang/Access.java (hunk 9 above)
Comment (ChengJin01) [reply jcl/src/java.base/share/classes/java/lang/Access.java:592]: FYI: @JasonFengJ9

File: runtime/jcl/common/thread.cpp (hunk 11 above)
Comment (ChengJin01) [reply runtime/jcl/common/thread.cpp:592]: FYI: @JasonFengJ9

File: jcl/src/java.base/share/classes/java/lang/Access.java [hunk 12]
```diff
@@ -39,6 +39,9 @@
 import com.ibm.oti.reflect.TypeAnnotationParser;
 
 /*[IF Sidecar19-SE]*/
+/*[IF JAVA_SPEC_VERSION >= 20]*/
+import java.io.InputStream;
+/*[ENDIF] JAVA_SPEC_VERSION >= 20 */
```
[the remark this answers is not in the document]
Comment (ChengJin01) [reply jcl/src/java.base/share/classes/java/lang/Access.java:44]: FYI: @JasonFengJ9

File: runtime/jcl/common/thread.cpp [hunk 13]
```diff
@@ -571,4 +571,27 @@ Java_java_lang_Thread_registerNatives(JNIEnv *env, jclass clazz)
...
+ * This is expected to invoke JVM_EnsureMaterializedForStackWalk and ensure
+ * that the stackwalk code sees a materialized value.
+ * This might be a hotspot specific requirement, i.e., an empty
+ * implementation could be sufficient for OpenJ9.
+ *
+ * @param env instance of JNIEnv
+ * @param unusedClass
+ * @param obj the object to be checked
+ *
+ * @return void
```
[the remark this answers is not in the document]
Comment (ChengJin01) [reply runtime/jcl/common/thread.cpp:587]: FYI: @JasonFengJ9

File: jcl/src/java.base/share/classes/java/lang/Access.java (hunk 9 above)
Comment (JasonFengJ9) [reply jcl/src/java.base/share/classes/java/lang/Access.java:592]: Removed

File: runtime/jcl/common/thread.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (JasonFengJ9) [reply runtime/jcl/common/thread.cpp:587]: Removed

File: runtime/jcl/common/thread.cpp (hunk 11 above)
Comment (JasonFengJ9) [reply runtime/jcl/common/thread.cpp:592]: Updated.

File: jcl/src/java.base/share/classes/java/lang/Access.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (JasonFengJ9) [reply jcl/src/java.base/share/classes/java/lang/Access.java:44]: Moved `initialSystemIn` into `System`, also assigned its value within `completeInitialization()` instead of `setIn()` which is a public method and could be invoked by an application.

File: runtime/jcl/common/thread.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (ChengJin01) [reply runtime/jcl/common/thread.cpp]: Picked up Jason's changes at https://github.com/eclipse-openj9/openj9/pull/16574.

File: runtime/jcl/common/thread.cpp (hunk 11 above)
Comment (ChengJin01) [reply runtime/jcl/common/thread.cpp]: Picked up Jason's changes at https://github.com/eclipse-openj9/openj9/pull/16574.

File: jcl/src/java.base/share/classes/java/lang/Access.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (ChengJin01) [reply jcl/src/java.base/share/classes/java/lang/Access.java:44]: Picked up Jason's changes at https://github.com/eclipse-openj9/openj9/pull/16574.

File: jcl/src/java.base/share/classes/java/lang/Access.java (hunk 9 above)
Comment (ChengJin01) [reply jcl/src/java.base/share/classes/java/lang/Access.java]: Picked up Jason's changes at https://github.com/eclipse-openj9/openj9/pull/16574.

**Comment by pshipton** [comment]
@keithc-ca did you get all the changes you requested?

File: jcl/src/java.base/share/classes/java/lang/System.java [hunk 14]
```diff
@@ -79,6 +79,10 @@ public final class System {
 	 * Default input stream
 	 */
 	public static final InputStream in = null;
+/*[IF JAVA_SPEC_VERSION >= 20]*/
+	static InputStream initialSystemIn;
```
[the remark this answers is not in the document]
Comment (ChengJin01) [reply jcl/src/java.base/share/classes/java/lang/System.java:83]: FYI: @JasonFengJ9

[the remark this answers is not in the document]
Comment (JasonFengJ9) [reply jcl/src/java.base/share/classes/java/lang/System.java:83]: Updated https://github.com/eclipse-openj9/openj9/pull/16574
@ChengJin01 pls pull it in

[the remark this answers is not in the document]
Comment (ChengJin01) [reply jcl/src/java.base/share/classes/java/lang/System.java]: Updated on top of the latest changes in Jason's branch.

File: jcl/src/java.base/share/classes/java/lang/System.java [hunk 15]
```diff
@@ -473,7 +478,11 @@ static void completeInitialization() {
 	/*[ENDIF]*/ // Sidecar18-SE-OpenJ9
 
 	/*[IF (Sidecar18-SE-OpenJ9|Sidecar19-SE)&!(PLATFORM-mz31|PLATFORM-mz64)]*/
-	setIn(new BufferedInputStream(new FileInputStream(FileDescriptor.in)));
+	InputStream tempIn = new BufferedInputStream(new FileInputStream(FileDescriptor.in));
+	setIn(tempIn);
+/*[IF JAVA_SPEC_VERSION >= 20]*/
+	initialIn = tempIn;
```
Comment (keithc-ca) [inline jcl/src/java.base/share/classes/java/lang/System.java:484]: This is redundant: the assignment to `initialIn` has already happened within `setIn()`.
Did you intend to remove the addition to `setIn()`?

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: jcl/src/java.base/share/classes/java/lang/System.java (hunk 15 above)
Comment (ChengJin01) [reply jcl/src/java.base/share/classes/java/lang/System.java:484]: FYI: @JasonFengJ9

Comment (JasonFengJ9) [reply jcl/src/java.base/share/classes/java/lang/System.java:484]: https://github.com/eclipse-openj9/openj9/blob/82d8e4396b1c29eb28bf38adec66cda4ec1b34a5/jcl/src/java.base/share/classes/java/lang/System.java#L547
`setIn()` is a public API, the `in` can be modified by the application.
`Access.initialSystemIn()` returns the first one set by bootstrap code.

Comment (pshipton) [reply jcl/src/java.base/share/classes/java/lang/System.java:484]: @keithc-ca the assignment is not done in setIn() (any more), are you looking at some old code?

Comment (keithc-ca) [reply jcl/src/java.base/share/classes/java/lang/System.java:484]: Perhaps I was.

File: jcl/src/java.base/share/classes/java/lang/System.java [hunk 16]
```diff
@@ -473,7 +478,11 @@ static void completeInitialization() {
...
+/*[ENDIF] JAVA_SPEC_VERSION >= 20 */
```
Comment (keithc-ca) [inline jcl/src/java.base/share/classes/java/lang/System.java:485]: Please indent the preprocessor comments to match nearby instances (e.g. lines 488, 494-498).

Comment (ChengJin01) [reply jcl/src/java.base/share/classes/java/lang/System.java]: Updated in my commit.
FYI: @JasonFengJ9

**Review by keithc-ca — APPROVED** [review]

**Comment by keithc-ca** [comment]
Built successfully with https://github.com/ibmruntimes/openj9-openjdk-jdk20/pull/12.
